### PR TITLE
core: remove the callbacks from kpatch_apply_patch and kpatch_remove_patch

### PR DESCRIPTION
**Pull request by zhenjiangma**
We encountered a problem refered to #1077 and  #1120. After analysis, we believe that the main reason is that the callbacks are placed in the context of stop_machine. Because there is no data-dependy, so we remove the callbacks from the functions of kpatch_apply_patch and kpatch_remove_patch. We will avoid being trapped into a dead lock while a sleep appears in callbacks and exactly a process sending IPI was scheduled.

### CODE REVIEW

**Comment by joe-lawrence** [comment]
@euspectre - if this looks interesting to support for kpatch.ko, do you mind reviewing?  Thanks.

**Comment by euspectre** [comment]
@joe-lawrence Yes, I'll take a look.

File: kmod/core/core.c [hunk 1]
```diff
@@ -742,7 +705,7 @@ static int kpatch_write_relocations(struct kpatch_module *kpmod,
 			break;
 		case R_X86_64_64:
 			loc = dynrela->dest;
```
Comment (euspectre) [inline kmod/core/core.c]: This part of the patch is unrelated. Looks like an artefact from rebase or something.
Please rebase the patch on top of the current master branch in dynup/kpatch.

**Comment by euspectre** [comment]
@zhenjiangma :  Commit https://github.com/dynup/kpatch/pull/1126/commits/f912b4aea2913a8c447447838587c4f2fb5853e6 is included into the pull requst but is not related to this.

As for "remove the callbacks from kpatch_apply_patch and kpatch_remove_patch" (https://github.com/dynup/kpatch/pull/1126/commits/45a8fca17f1baf141146eb5802d99ca15eba2df9) - I cannot see the description in the commit. The change is non-trivial, so it should be described in the commit message too. Otherwise, even the author would have a hard time, in a few months, remembering why these changes were needed.

-------

As far as I can see, you moved all four kinds of callbacks out of stop_machine(). As I suggested in https://github.com/dynup/kpatch/issues/1077, it is probably safe for pre-patch and post-unpatch callbacks because no races with the patched code are possible there. Should be enough for the most cases, including https://github.com/dynup/kpatch/issues/1120.

I'd leave post-patch and pre-unpatch callbacks under stop_machine() for now, because these could run in parallel with the patched code otherwise, leading to potential bugs.

File: kmod/core/core.c (hunk 1 above)
Comment (zhenjiangma) [reply kmod/core/core.c]: Yes, maybe I got an old version. I will rebase the patch on top of the current master branch in dynup/kpatch.

**Comment by zhenjiangma** [comment]
> @zhenjiangma : Commit [f912b4a](https://github.com/dynup/kpatch/commit/f912b4aea2913a8c447447838587c4f2fb5853e6) is included into the pull requst but is not related to this.
> 
> As for "remove the callbacks from kpatch_apply_patch and kpatch_remove_patch" ([45a8fca](https://github.com/dynup/kpatch/commit/45a8fca17f1baf141146eb5802d99ca15eba2df9)) - I cannot see the description in the commit. The change is non-trivial, so it should be described in the commit message too. Otherwise, even the author would have a hard time, in a few months, remembering why these changes were needed.
> 
> As far as I can see, you moved all four kinds of callbacks out of stop_machine(). As I suggested in #1077, it is probably safe for pre-patch and post-unpatch callbacks because no races with the patched code are possible there. Should be enough for the most cases, including #1120.
> 
> I'd leave post-patch and pre-unpatch callbacks under stop_machine() for now, because these could run in parallel with the patched code otherwise, leading to potential bugs.

Yes, as you mentioned above, pre-patch and post-unpatch callbacks are safe.  I didn't considered the race between the patched code and post-patch/pre-unpatch callbacks when run in parallel. In action,  we use pre_patch and post_unpatch usually,  and it is enough for most cases. I will repull it as you suggested above.

**Comment by zhenjiangma** [comment]
> @zhenjiangma : Commit [f912b4a](https://github.com/dynup/kpatch/commit/f912b4aea2913a8c447447838587c4f2fb5853e6) is included into the pull requst but is not related to this.
> 
> As for "remove the callbacks from kpatch_apply_patch and kpatch_remove_patch" ([45a8fca](https://github.com/dynup/kpatch/commit/45a8fca17f1baf141146eb5802d99ca15eba2df9)) - I cannot see the description in the commit. The change is non-trivial, so it should be described in the commit message too. Otherwise, even the author would have a hard time, in a few months, remembering why these changes were needed.
> 
> As far as I can see, you moved all four kinds of callbacks out of stop_machine(). As I suggested in #1077, it is probably safe for pre-patch and post-unpatch callbacks because no races with the patched code are possible there. Should be enough for the most cases, including #1120.
> 
> I'd leave post-patch and pre-unpatch callbacks under stop_machine() for now, because these could run in parallel with the patched code otherwise, leading to potential bugs.

I pull a new request #1134 to make modification according to your suggestion, because I don't know how to recommit on this request.

**Comment by euspectre** [comment]
> I pull a new request #1134 to make modification according to your suggestion, because I don't know how to recommit on this request.

You could, for example, drop the old commit(s) from your branch (see 'git rebase -i'), add the new commits and force-push them here (git push -f <...>).

If you authored the patch entirely in GitHub's web UI (not sure if that is possible), then I also don't know how to replace it with a new one.

Anyway, I'll take a look at the new patch.

**Pull request closed without merging**